### ISSUE TITLE
Automatic revocation of access tokens causes trouble for multi process applications

### ISSUE THREAD

**Issue by NobodysNightmare** [doorkeeper-gem/doorkeeper#1314]
### Meta

I am making an issue from a [pull request discussion](https://github.com/doorkeeper-gem/doorkeeper/pull/1271#discussion_r318169687) on PR #1271.

### Description

PR #1271 introduced automatic revocation of access tokens, when the same client already fetched an access token.

The problem is that multiple processes of the same client may request and cache tokens in their local memory. Subsequent requests to get tokens by other processes can end up revoking an in use token that another process has cached.

### Example scenario

1. Process A fetches token A and stores it in memory
1. Process B fetches a token
1. `doorkeeper` determines to not send the existing token A (either reuse_access_tokens is off or the token_reuse_limit has been reached)
1. **Server revokes token A**
1. Server generates and returns token B
1. Process B caches token B in memory
1. Process A **attempts to use token A** (it was unaware token A was revoked)

### Expected behavior
An application split across multiple processes/servers does not need shared state to handle access tokens.

Disclaimer: I did not yet cross-check the OAuth 2 RFC on whether clients should be allowed to fetch multiple access tokens, but I am pretty sure it is at least not explicitly prohibited.

### Actual behavior
Two processes of the same client can inadvertently revoke each others cached access tokens.

### Compatibility note

I would consider this a breaking (or at least **[IMPORTANT]**) change, that should be mentioned in the CHANGELOG. So depending of the outcome of this issue, we should either note a breaking change starting in 5.2 or at least one introduced there and fixed later.

**Comment by NobodysNightmare** [comment]
I just thought about my last proposal further. A simplification might be to just adapt [`authorized_tokens_for`](https://github.com/doorkeeper-gem/doorkeeper/blob/b69df59708b2d5aaa36fdbb331552d9dd53b8cab/lib/doorkeeper/models/access_token_mixin.rb#L193-L198) to _filter_ by the `created_at`, taking the TTL into account.

This should reduce the number of returned tokens tremendously and works without a change to the schema. Example implementation:

```ruby
def authorized_tokens_for(application_id, resource_owner_id)
  ordered_by(:created_at, :desc)
    .where(application_id: application_id,
           resource_owner_id: resource_owner_id,
           revoked_at: nil)
    .where('created_at > ?', Doorkeeper.configuration.access_token_expires_in.seconds.ago)
end
```

I am keeping my prior proposals, in case they trigger some good thoughts for anyone.

**Comment by nbulaj** [comment]
@NobodysNightmare I like the idea with time math using database if it's possible and fit's all the possible configuration cases (I remember we have some custom expiration time configurations that must be considered).

**Comment by drock** [comment]
I left the PR comment that originally spurred this issue.  When I discovered this issue with my team, we discussed paths forward and the time based query approach was exactly the idea we had as well.

**Comment by nbulaj** [comment]
So do you @NobodysNightmare have any plans to propose a enhancement PR or c PoC with the time math described above?

**Comment by NobodysNightmare** [comment]
Sorry for not following up here. I created a draft-like PR now: https://github.com/doorkeeper-gem/doorkeeper/pull/1318

Let me know, whether it shoots in the right direction.

**Comment by nbulaj** [comment]
New option and regression fix introduced here #1353 and would be released with Doorkeeper 5.3

**Comment by nbulaj** [comment]
Hi @NobodysNightmare . #1353 releases with a new version so feel free to close this issue if it's no more valid
